src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -750,5 +733,311 @@ private static string ToHtmlCase(string name)
+                var nameStartIndex = _index;
+                var nameEndIndex = _requiredAttributes.IndexOfAny(new char[] { ' ', '\t', ']', '=', '^', '$' }, _index);
+                nameEndIndex = nameEndIndex == -1 ? _requiredAttributes.Length : nameEndIndex;
+                _index = nameEndIndex;
+
+                var attributeName = _requiredAttributes.Substring(nameStartIndex, nameEndIndex - nameStartIndex);
+
+                return attributeName;
+            }
+
+            private char ParseCSSValueOperator(ErrorSink errorSink)
+            {
+                Debug.Assert(!AtEnd);
+
+                if (TagHelperRequiredAttributeDescriptor.SupportedCSSValueOperators.Contains(Current))
+                {
+                    var op = Current;
+                    Next();
+
+                    if (op != '=' && At('='))
+                    {
+                        // Two length operator (ex: ^=). Move past the second piece
+                        Next();
+                    }
+
+                    return op;
+                }
+
+                return '\0';
+            }
+
+            private string ParseCSSValue(ErrorSink errorSink)
+            {
+                int valueStart, valueEnd;
+                if (At('\'') || At('"'))
+                {
+                    var quote = Current;
+
+                    // Move past the quote
+                    Next();
+
+                    valueStart = _index;
+                    var quoteEndIndex = _requiredAttributes.IndexOf(quote, _index);
+                    if (quoteEndIndex == -1)
+                    {
+                        errorSink.OnError(
+                            SourceLocation.Zero,
+                            Resources.FormatTagHelperDescriptorFactory_InvalidRequiredAttributeCharacterExpectedOther(
+                                _requiredAttributes[_requiredAttributes.Length - 1],
+                                quote),
+                            length: 0);
+                        return null;
+                    }
+                    valueEnd = quoteEndIndex;
+                    _index = quoteEndIndex + 1;
+                }
+                else
+                {
+                    valueStart = _index;
+                    var valueEndIndex = _requiredAttributes.IndexOfAny(InvalidCSSQuotelessValueCharacters, _index);
+                    valueEnd = valueEndIndex == -1 ? _requiredAttributes.Length : valueEndIndex;
+                    _index = valueEnd;
+                }
+
+                var value = _requiredAttributes.Substring(valueStart, valueEnd - valueStart);
+
+                return value;
+            }
+
+            private TagHelperRequiredAttributeDescriptor ParseCSSSelector(ErrorSink errorSink)
+            {
+                Debug.Assert(At('['));
+
+                // Move past '['.
+                Next();
+                PassOptionalWhitespace();
+
+                var attributeName = ParseCSSAttributeName(errorSink);
```

Comment:
Won't this `throw` if string ends with `"["` or `"[    "`? Again, might be useful for `PassOptionalWhitespace()` to return a bool indicating there's more to come. Might be even better for that method to be passed the `ErrorSink` and a `bool` indicating that reaching the end should result in an `...UnexpectedEnd...` error.